### PR TITLE
fix: Don't kill supervisord when processes exit, for #5906, fixes #5358

### LINKED ISSUE #5358: supervisord should not bring down the whole web container when non-essential service fails
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Output of `ddev debug test`

<details><summary>Expand `ddev debug test` diagnostic information</summary>

```
Running bash [-c /var/folders/k2/l6n0dl6j4dg_8x02bwvjfwrr0000gn/T/test_ddev.sh] 
======= Existing project config =========
These config files were loaded for project my-project: [/Users/yohanescandrajaya/Documents/GitHub/ddev/my-project/.ddev/config.yaml /Users/yohanescandrajaya/Documents/GitHub/ddev/my-project/.ddev/config.development.yaml /Users/yohanescandrajaya/Documents/GitHub/ddev/my-project/.ddev/config.local.yaml /Users/yohanescandrajaya/Documents/GitHub/ddev/my-project/.ddev/config.selenium-standalone-chrome.yaml]
name: my-project
type: php
docroot: web
php_version: 8.2
webserver_type: apache-fpm
webimage: ddev/ddev-webserver:20230803_php_serialize_precision
xdebug_enabled: true
database: {mariadb 10.11}
performance_mode: none
hooks: map[post-start:[map[exec-host:composer config github-oauth.github.com 2>/dev/null | xargs -I {} ddev exec "composer config github-oauth.github.com {}"] map[exec:sh d/initLocal.sh] map[exec:sh d/migrate.sh] map[exec-host:composer config github-oauth.github.com 2>/dev/null | xargs -I {} ddev exec "composer config github-oauth.github.com {}"] map[exec:sh d/initLocal.sh] map[exec:sh d/migrate.sh]] pre-start:[map[exec-host:[ ! -f .ddev/docker-compose.selenium-chrome.yaml ] || rm .ddev/docker-compose.selenium-chrome.yaml] map[exec-host:[ ! -f .ddev/docker-compose.selenium-firefox.yaml ] || rm .ddev/docker-compose.selenium-firefox.yaml] map[exec-host:sh fetchDeploy.sh] map[exec-host:[ ! -f .ddev/docker-compose.selenium-chrome.yaml ] || rm .ddev/docker-compose.selenium-chrome.yaml] map[exec-host:[ ! -f .ddev/docker-compose.selenium-firefox.yaml ] || rm .ddev/docker-compose.selenium-firefox.yaml] map[exec-host:sh fetchDeploy.sh]]]
upload_dirs: [/var/www/html/storage]
host_db_port: 59099
mailhog_port: 8025
mailhog_https_port: 8026
project_tld: ddev.site
use_dns_when_possible: true
composer_version: 2
nodejs_version: 18
default_container_timeout: 120
web_extra_exposed_ports: [{web 8088 8088 8089}]
======= Creating dummy project named  tryddevproject-25810 in ../tryddevproject-25810 =========
```
</details>         

### Expected Behavior

When a command that runs by the Supervisor gets an error or crashes, it automatically restarts the app with autorestart configuration on the supervisor.

### Actual Behavior

When a command that runs by the Supervisor gets an error or crash, it automatically turns off the web container. Making the supervisor auto restart feature useless.

### Steps To Reproduce

1. create a daemon that runs by supervisord
2. ddev start
3. make the daemon crash by making an error/throw an exception
4. see in the docker dashboard that the web container is shut down

### Anything else?

I think it's related to this issue here
https://github.com/ddev/ddev/issues/1137

and this commit
https://github.com/ddev/ddev/commit/58f80232c65a961c935f85417a92c2df310a5f34

I could make a turnaround by replacing the supervisor.conf without the `[eventlistener:child_exit_monitor]` section. but supervisor behavior should work as expected out of the box with ddev since I've spent some hours getting to this turnaround. This issue does not exist in the documentation as well.

### PR DESCRIPTION
**Pull request by stasadev**
## The Issue

- #5358
- #5906
- #1142

## How This PR Solves The Issue

The logic added in #1142 was done before healthchecks were added to Docker containers.

The change proposed in #5906 makes the logic even more complex.

Alternatively, we can remove `kill_supervisor.py` completely.

## Manual Testing Instructions

Use this in `.ddev/config.yaml`:

```
web_extra_daemons:
  - name: baddaemon
    command: somemissingcommand
    directory: /var/www/html
  - name: gooddaemon
    command: 'while true; do date; sleep 60; done'
    directory: /var/www/html
```

```
ddev logs # see correct behavior for baddaemon
WARN exited: baddaemon (exit status 127; not expected)
```

```
ddev exec killall php-fpm
ddev logs # see it restarting again
INFO success: php-fpm entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

### CODE REVIEW

**Comment by rfay** [comment]
This is working, and I think it's an improvement. However, we have a bit more work to do to make it really work right.

- [ ] Docs explaining how you see if a daemon is working (`ddev logs | grep INFO` for example, or `ddev logs | grep <daemonname>`)
- [ ] I'm pretty sure we have to do something with `exitcodes` (http://supervisord.org/configuration.html#program-x-section-settings) in the generated supervisord config; right now, a failure does *not* stop the restart behavior, the the output is not correct. 

I used this configuration in config.yaml:
```
web_extra_daemons:
  - name: baddaemon
    command: somemissingcommand
    directory: /var/www/html
  - name: gooddaemon
    command: 'while true; do date; sleep 60; done'
    directory: /var/www/html
```

It generates a config like this for baddaemon:
```
[program:baddaemon]
group=webextradaemons
command=bash -c "somemissingcommand || sleep 2"
directory=/var/www/html
autostart=false
autorestart=true
startretries=15
stdout_logfile=/var/tmp/logpipe
stdout_logfile_maxbytes=0
redirect_stderr=true
```

and then `ddev start` and `ddev logs`. I see:

```
2024-03-05 21:39:01,009 INFO success: mailpit entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:01,494 INFO spawned: 'baddaemon' with pid 2190
2024-03-05 21:39:01,495 INFO spawned: 'gooddaemon' with pid 2191
bash: line 1: somemissingcommand: command not found
Tue Mar  5 21:39:01 UTC 2024
2024-03-05 21:39:02,524 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:02,525 INFO success: gooddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:03,523 INFO exited: baddaemon (exit status 0; expected)
2024-03-05 21:39:04,533 INFO spawned: 'baddaemon' with pid 2296
bash: line 1: somemissingcommand: command not found
2024-03-05 21:39:05,578 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:06,576 INFO exited: baddaemon (exit status 0; expected)
2024-03-05 21:39:07,582 INFO spawned: 'baddaemon' with pid 2327
bash: line 1: somemissingcommand: command not found
2024-03-05 21:39:08,624 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

We see that `baddaemon` is of course failing, and *not* staying up for more than 1 second. (Probably we need `startretries` > 1, perhaps 3?) And supervisord is saying that it *is* staying up, which isn't true.

**Comment by rfay** [comment]
👍 Also, `ddev exec killall mailpit` kills mailpit and it gets restarted
However, apache, nginx, and php-fpm do not restart, and they'd be better that way, `autorestart=true`

**Comment by stasadev** [comment]
I repushed the Docker tag:

```
docker pull ddev/ddev-webserver:20240305_stasadev_remove_kill_supervisor
```

> Docs explaining how you see if a daemon is working (`ddev logs | grep INFO` for example, or `ddev logs | grep <daemonname>`)

It's better to check the full output, and it's already described in the documentation, but it's not as noticeable as it should be, I made a small change for this.

> right now, a failure does not stop the restart behavior, the the output is not correct.
> We see that `baddaemon` is of course failing, and not staying up for more than 1 second. (Probably we need `startretries` > 1, perhaps 3?) And supervisord is saying that it is staying up, which isn't true.

We already had `startretries=15`, the problem was in `command=bash -c "somemissingcommand || sleep 2"`, `sleep` always returned 0.

I found the reason for `sleep 2`: [Use bash for web_extra_daemons command and add a sleep to wait for mutagen](https://github.com/ddev/ddev/pull/4057/commits/491c10203d9a9ac00a37d1aed69e35a07021e0cf).

I refactored it so that the command's return code was always returned, and added `startsecs=3` to wait 3 secs (sleep 2 + 1 extra sec) for success.

File: pkg/ddevapp/config.go [hunk 1]
```diff
@@ -948,10 +948,11 @@ func (app *DdevApp) RenderComposeYAML() (string, error) {
 		supervisorConf := fmt.Sprintf(`
 [program:%s]
 group=webextradaemons
-command=bash -c "%s || sleep 2"
+command=bash -c "%s; exit_code=$?; if [ $exit_code -ne 0 ]; then sleep 2; fi; exit $exit_code"
```
Comment (rfay) [inline pkg/ddevapp/config.go:951]: Ah, this is also why it was reported as staying up more than 1s

**Review by rfay — APPROVED** [review]
Well, I really like it. I think it's more useful behavior all the way around, although it is definitely a change. 

Tested with the manual instructions and with `killall php-fpm` and `killall nginx` and everything came back and started working again.

**Comment by rfay** [comment]
Rebased and pushed new image to pick up other image changes. Now with the test I see
```
Starting web_extra_daemons...
Unable to start web_extra_daemons using supervisorctl, stdout=webextradaemons:baddaemon: ERROR (spawn error)
webextradaemons:gooddaemon: started
, stderr=: composeCmd failed to run 'COMPOSE_PROJECT_NAME=ddev-d10 docker-compose -f /Users/rfay/workspace/d10/.ddev/.ddev-docker-compose-full.yaml exec -T web bash -c set -eu && ( supervisorctl start webextradaemons:*)', action='[]', err='exit status 7', stdout='webextradaemons:baddaemon: ERROR (spawn error)
webextradaemons:gooddaemon: started
', stderr=''
```

Which actually seems like a pretty good thing, but I don't think it's what I had before.

**Comment by rfay** [comment]
It looks to me like this is all great, so pulling. 

To test, make sure to re-pull the latest 

```
docker pull ddev/ddev-webserver:20240305_stasadev_remove_kill_supervisor
```

**Pull request merged**